### DISCUSSION TITLE
CUDA not using the GPU

### DISCUSSION THREAD

**Discussion by veczieds** [trackmania-rl/tmrl#43] [Q&A] [answered]
Hello!
I set up a virtual machine with a Tesla T4 GPU so i can do the training on that machine. But i cant make the training utilize the GPU. In the config file CUDA is set to true. I even tried running competition script with "cuda:0" listed as the device to use for training but nothing. All the drivers seem to be configured correctly. I even tried to use WDDM and TCC drivers but nothing with either of them. nvidia-smi doesnt show any python processes running on the card. Maybe someone could point where to look.
Just quickly wanted to add that this project is amazing. Im not too familiar with machine learning, but project creators have made it easy to get into this stuff.

**Comment by yannbouteiller** [comment]
Hi, thanks! :)

Probably your issue comes from the fact you are using a virtual machine?

On my laptop, running both the worker (with trackmania) and the trainer (with CUDA enabled), the trainer uses CUDA alright:

![smi tmrl](https://user-images.githubusercontent.com/15182559/230167757-bf117d8f-7136-4b71-9bca-8c9f346762fc.png)

(this is Windows but I usually have the trainer run remotely on a Linux machine)

Probably you didn't miss that, but just to double-check, in the `config.json` file that is on your trainer machine you want something like this:
```json
"CUDA_TRAINING": true,
"CUDA_INFERENCE": false,
```

(Inference is for the worker, training is for the trainer)

Maybe on your virtual machine you can start by double-checking that pytorch finds CUDA:
```terminal
(tmrl) C:\Users\yannb>python
Python 3.10.8 | packaged by conda-forge | (main, Nov 24 2022, 14:07:00) [MSC v.1916 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.cuda.is_available()
True
```

**Reply by veczieds** [reply]
I did some more reading and already checked if pytorch finds CUDA. And it doesnt. Im trying to look for a reason why it cant find CUDA as i have the compatible versions. At the moment i installed CUDA 11.8 cause Pytorch website showed pytorch installation for 11.8.

**Reply by yannbouteiller** [reply, marked as answer]
Yeah that's you issue then. If you can, I recommend using anaconda and their CUDA version of pytorch, this usually fixes this kind of issues (although not sure on a VM):
```terminal
conda install pytorch torchvision torchaudio pytorch-cuda=11.8 -c pytorch -c nvidia
```

**Reply by veczieds** [reply]
I avoided using anaconda. Will try this and come back with results

**Reply by veczieds** [reply]
Looks like it solved the problem. Thank you very much!

**Reply by yannbouteiller** [reply]
Great! You're welcome :)
